### PR TITLE
[SPARK-38914][SQL] Allow user to insert specified columns into insertable view

### PR DESCRIPTION
**Pull request by morvenhuang**
### What changes were proposed in this pull request?
Allow user to insert specified columns into insertable view, for example, 
```
CREATE TEMPORARY VIEW v1 (c1 int, c2 string) USING org.apache.spark.sql.json.DefaultSource OPTIONS (path 'json_dir')
INSERT INTO v1(c1) VALUES(100)
SELECT c1, c2 FROM v1;
+---+----+
| c1|  c2|
+---+----+
|100|null|
+---+----+
```

### Why are the changes needed?
The option spark.sql.defaultColumn.useNullsForMissingDefautValues allows us to insert specified columns into table (https://issues.apache.org/jira/browse/SPARK-38795), but currently this option does not work for insertable view, 

To keep consistenct with the behavior of INSERT INTO table, we should also allow user to specify columns when running INSERT INTO view.

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
New unit tests added.

### CODE REVIEW

**Comment by morvenhuang** [comment]
@dtenedor I made a bit code change on class ResolveDefaultColumns, could you please help to take a look when you have time? Many thanks.

**Comment by morvenhuang** [comment]
@HyukjinKwon Hi Hyukjin, could you please be kind enough to verify this? And also do you know why the workflow 'Build and test' failed with 'workflow run detection failed', seems something wrong with the github build?

**Comment by dtenedor** [comment]
@morvenhuang apologies for the delay in review, I was on vacation for the last few days :)

The changes in `ResolveDefaultColumns` look like they are on the right track. I am thinking, now we have this SQLConf to add NULL values for INSERTs that specify only a subset of the columns, but it is disabled by default. If we want to add functionality based on this, and we think it is useful, let's just enable it by default in this PR. This is safe because it only changes INSERTs that would have returned errors to execute successfully instead. We would need to update the affected tests; this could be done without much work by just pushing a commit setting it to true by default and then let the automated tests run. Then the subset of tests that subsequently fail should indicate what needs to be changed, and the change should be relatively straightforward.

**Comment by morvenhuang** [comment]
@dtenedor oh, thank you for the reply, how was the vacation? 
Yeah, I agree with you that it's better to enable the option by default, I'm working on it.

BTW, it seems now we cannot trigger 'build and test' by pushing a commit, always get 'workflow run detection failed' error. Do you have any idea what the cause is, and how to get around the problem? Many thanks.

**Comment by dtenedor** [comment]
@morvenhuang thanks, it was fun! Took three days off, it was very relaxing.

The `workflow run detection failed` error is due to a current Spark-wide breakage in the REST API for the Github checks. Looking at the set of open PRs, many of them show this problem:

https://github.com/apache/spark/pulls

I imagine it should resolve once the core issue is fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 1]
```diff
@@ -280,6 +280,9 @@ case class ResolveDefaultColumns(
       case SubqueryAlias(_, r: UnresolvedCatalogRelation) =>
         StructType(r.tableMeta.schema.fields.dropRight(
           enclosingInsert.get.partitionSpec.size))
+      case SubqueryAlias(_, r: View) =>
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:283]: I'm a bit confused about what it means to `INSERT INTO` a view. Can you leave a comment

File: sql/core/src/test/scala/org/apache/spark/sql/connector/InsertIntoTests.scala [hunk 2]
```diff
@@ -118,16 +118,18 @@ abstract class InsertIntoTests(
   }
 
   test("insertInto: fails when missing a column") {
-    val t1 = s"${catalogAndNamespace}tbl"
-    sql(s"CREATE TABLE $t1 (id bigint, data string, missing string) USING $v2Format")
-    val df = Seq((1L, "a"), (2L, "b"), (3L, "c")).toDF("id", "data")
-    val exc = intercept[AnalysisException] {
-      doInsert(t1, df)
-    }
+    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (dtenedor) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/InsertIntoTests.scala:121]: Could we also add a version of this test with the conf set to "true" to show that the intended behavior works? (This should be the same as the test code in this case before this PR.)

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 3]
```diff
@@ -144,14 +144,16 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
   test("SELECT clause generating a different number of columns is not allowed.") {
-    val message = intercept[AnalysisException] {
-      sql(
-        s"""
-        |INSERT OVERWRITE TABLE jsonTable SELECT a FROM jt
-      """.stripMargin)
-    }.getMessage
-    assert(message.contains("target table has 2 column(s) but the inserted data has 1 column(s)")
-    )
+    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (dtenedor) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:147]: Same as above, could we also add a version of this test with the conf set to "true" to show that the intended behavior works? (This should be the same as the test code in this case before this PR.)

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 4]
```diff
@@ -1078,11 +1101,13 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
       }
     }
     // There is one trailing default value referenced implicitly by the INSERT INTO statement.
-    withTable("t") {
-      sql("create table t(i int, s bigint default 42, x bigint) using parquet")
-      assert(intercept[AnalysisException] {
-        sql("insert into t values(1)")
-      }.getMessage.contains("expected 3 columns but found"))
+    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (dtenedor) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1104]: Same as above, could we also add a version of this test with the conf set to "true" to show that the intended behavior works? (This should be the same as the test code in this case before this PR.)

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 5]
```diff
@@ -1178,34 +1205,34 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
-    withTable("t") {
-      sql("create table t(i boolean, s bigint, q int default 43) using parquet")
-      assert(intercept[AnalysisException] {
-        sql("insert into t (i, q) select true from (select 1)")
-      }.getMessage.contains(addTwoColButExpectedThree))
-    }
-    // When the USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES configuration is disabled, and no
-    // explicit DEFAULT value is available when the INSERT INTO statement provides fewer
-    // values than expected, the INSERT INTO command fails to execute.
     withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (dtenedor) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1208]: Same as above, could we also add a version of this test with the conf set to "true" to show that the intended behavior works?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 6]
```diff
@@ -1467,10 +1494,12 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
   }
 
   test("SPARK-36980: Insert support query with CTE") {
-    withTable("t") {
-      sql("CREATE TABLE t(i int, part1 int, part2 int) using parquet")
-      sql("INSERT INTO t WITH v1(c1) as (values (1)) select 1, 2, 3 from v1")
-      checkAnswer(spark.table("t"), Row(1, 2, 3))
+    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (dtenedor) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1497]: Same as above, could we also add a version of this test with the conf set to "true" to show that the intended behavior works? (This should be the same as the test code in this case before this PR.)

**Review by dtenedor — CHANGES_REQUESTED** [review]
This mostly LG, we probably just want to update the unit tests so they cover behavior with the config both enabled and disabled.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 4 above)
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1104]: Daniel, I see several tests in test("INSERT INTO statements with tables with default columns: positive tests") already cover this scenario.

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1104]: OK, if there are existing tests that cover it when the `SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES` is set to true, then no need to add more tests for those cases then.

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1104]: (Please feel free to resolve any other code-review comments where that is the case.)

File: sql/core/src/test/scala/org/apache/spark/sql/connector/InsertIntoTests.scala (hunk 2 above)
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/InsertIntoTests.scala:122]: New test("insertInto: use null for the missing column") added.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 3 above)
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:147]: New test("SELECT clause generating a different number of columns: positive test.") added.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 5 above)
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1221]: There is an existing test("INSERT INTO with user specified columns and defaults: positive tests") for this.

**Review by dtenedor — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]
Build link: https://github.com/morvenhuang/spark/runs/6071598807

Should be good to go. cc @gengliangwang @cloud-fan FYI.

**Comment by morvenhuang** [comment]
Thank you guys for the review, I've made some more changes on tests, now the build&test should be OK.

@dtenedor Daniel, I realize there're some minor issues, one is that `DSV2SQLInsertTestSuite: test("insert with column list - mismatched column list size")` failed when `useNullsForMissingDefaultValues=true`, this seems due to some catalog problem. 
The other one is `test("SPARK-36980: Insert support query with CTE")`, this test also failed when `useNullsForMissingDefaultValues=true`.
These two look pretty tricky to me, and still cannot figure out how to fix them, so probably you can help to create some new PRs for them?

**Comment by dtenedor** [comment]
> @dtenedor Daniel, I realize there're some minor issues, one is that DSV2SQLInsertTestSuite: test("insert with column list - mismatched column list size") failed when useNullsForMissingDefaultValues=true, this seems due to some catalog problem.
The other one is InsertSuite: test("SPARK-36980: Insert support query with CTE"), this test also failed when useNullsForMissingDefaultValues=true.
These two look pretty tricky to me, and still cannot figure out how to fix them, so probably we can just merge the current PR, and then you can help to create some new PRs for these issues, your thoughts?

Sure, I can just make a separate PR to update those tests with the useNullsForMissingDefaultValues set to both true and false in each test case, to verify the behavior.

**Comment by dtenedor** [comment]
@morvenhuang Note, I made a small update to `ResolveDefaultColumns` to fix the two tests you mentioned in [1]. Thanks for pointing that out, it gives us a chance to improve that code.

[1] https://github.com/apache/spark/pull/36122

**Comment by morvenhuang** [comment]
@dtenedor Thank you for the patience and the help.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -2858,15 +2858,15 @@ object SQLConf {
...
+    buildConf("spark.sql.defaultColumn.useNullsForMissingDefaultValues")
       .internal()
       .doc("When true, and DEFAULT columns are enabled, allow column definitions lacking " +
         "explicit default values to behave as if they had specified DEFAULT NULL instead. " +
         "For example, this allows most INSERT INTO statements to specify only a prefix of the " +
         "columns in the target table, and the remaining columns will receive NULL values.")
       .version("3.4.0")
       .booleanConf
-      .createWithDefault(false)
+      .createWithDefault(true)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2869]: This is a breaking change. Why do we have to make it `true`? (Probably I miss some discussions in this PR.)

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2869]: @gengliangwang No problem, per our discussion:
* The previous behavior when inserting fewer values than the number of columns in the table was to return an error, e.g. `CREATE TABLE t (a INT, b INT); INSERT INTO t VALUES 42'.
* When updating this `USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES` config to true, then such commands would instead succeed and insert NULL values for remaining columns in the table. This is the behavior that some other engines use, e.g. Postgres, which some users have requested in Spark.
* This would be a behavior change (which some users may not expect) but not technically a breaking change because it changes only commands that would have returned errors to completing successfully instead.

image

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2869]: Maybe we could leave the config as `false` in this PR to not block merging it, and discuss the change with others for a few days? Then if we decide to change it to true, we can do so in a separate PR, WDYT?

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2869]: Update on this: we talked a bit and decided it would be best to keep this config false by default. We can retain the test coverage but leave the existing OSS Spark behavior as-is. Thanks for discussing the details for this.

Comment (morvenhuang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2869]: OK, I'll change it back to false.

Comment (morvenhuang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Done. Code updated, build&test succeeded

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 8]
```diff
@@ -131,17 +131,21 @@ case class ResolveDefaultColumns(
...
-    val schema = insertTableSchemaWithoutPartitionColumns.getOrElse(return None)
-    val newDefaultExpressions: Seq[Expression] = getDefaultExpressions(numQueryOutputs, schema)
-    val newAliases: Seq[NamedExpression] =
-      newDefaultExpressions.zip(schema.fields).map {
-        case (expr, field) => Alias(expr, field.name)()
-      }
-    if (newDefaultExpressions.nonEmpty) {
-      Some(project.copy(projectList = project.projectList ++ newAliases))
-    } else {
+    if (project.projectList.exists(_.isInstanceOf[Star])) {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:134]: shall we move this check to line 82?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala [hunk 9]
```diff
@@ -183,35 +183,39 @@ trait SQLInsertTestSuite extends QueryTest with SQLTestUtils {
...
 
-    withTable("t1") {
-      createTable("t1", cols, Seq.fill(4)("int"))
-      val e1 = intercept[AnalysisException](sql(s"INSERT INTO t1 (c1) values(1)"))
-      assert(e1.getMessage.contains("target table has 4 column(s) but the inserted data has 1") ||
-        e1.getMessage.contains("expected 4 columns but found 1") ||
-        e1.getMessage.contains("not enough data columns") ||
-        e1.getMessage.contains(v2Msg))
-    }
+    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:200]: Why do we need this?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:200]: There are so many tests setting this conf. Shall we clean them up?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 10]
```diff
@@ -142,17 +142,30 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
   }
 
-  test("SELECT clause generating a different number of columns is not allowed.") {
+  test("SELECT clause generating a different number of columns: negative test.") {
     val message = intercept[AnalysisException] {
       sql(
         s"""
-        |INSERT OVERWRITE TABLE jsonTable SELECT a FROM jt
-      """.stripMargin)
+            |INSERT OVERWRITE TABLE jsonTable SELECT a FROM jt
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:149]: nit: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 11]
```diff
@@ -1632,6 +1662,24 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     }
   }
 
+  test("SELECT clause with star wildcard") {
+    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "true") {
+      withTable("t1") {
+        sql("CREATE TABLE t1(c1 int, c2 string) using parquet")
+        sql("INSERT INTO TABLE t1 select * from jt where a=1")
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1669]: We don't need to set SQLConf for this case.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 12]
```diff
@@ -443,6 +443,76 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
...
+        }
+
+        sql(s"INSERT INTO TABLE $tableName PARTITION (b=14, c) SELECT 13, 16, 15")
+
+        // Dynamic partitioning columns need to be after static partitioning columns.
+        intercept[AnalysisException] {
+          sql(s"INSERT INTO TABLE $tableName PARTITION (b, c=19) SELECT 17, 20, 18")
+        }
+
+        sql(s"INSERT INTO TABLE $tableName PARTITION (b, c) SELECT 17, 20, 18, 19")
```
Comment (gengliangwang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala:496]: shall we group all the insertions together?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 13]
```diff
@@ -443,6 +443,76 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
       }
   }
 
+  testPartitionedTable("INSERT INTO a partitioned table, use nulls for missing columns") {
```
Comment (gengliangwang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala:446]: This case doesn't seem to related to insertable view

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala (hunk 8 above)
Comment (morvenhuang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala]: Thank you for pointing out this, I've moved it to line 82.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala (hunk 9 above)
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala]: Done, all unnecessary USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES=false settings have been removed.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 14]
```diff
@@ -142,17 +142,30 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
-      """.stripMargin)
+            |INSERT OVERWRITE TABLE jsonTable SELECT a FROM jt
+         """.stripMargin)
     }.getMessage
     assert(message.contains("target table has 2 column(s) but the inserted data has 1 column(s)")
     )
   }
 
+  test("SELECT clause generating a different number of columns: positive test.") {
+    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "true") {
```
[the remark this answers is not in the document]
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: A little bit different, but you're right, I've move this to the "Allow user to insert specified columns into insertable view".

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 10 above)
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: All these accidental changes have been reverted.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 11 above)
Comment (morvenhuang) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala (hunk 13 above)
Comment (morvenhuang) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala]: Removed this test since it's not relevant.

**Comment by morvenhuang** [comment]
@gengliangwang thanks a lot for all your comments, please help to verify when you have time.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 15]
```diff
@@ -1200,43 +1225,41 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     // When the USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES configuration is disabled, and no
     // explicit DEFAULT value is available when the INSERT INTO statement provides fewer
     // values than expected, the INSERT INTO command fails to execute.
-    withSQLConf(SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1203]: Let's keep this `withSQLConf`, in case we change the default value in the future release.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala [hunk 16]
```diff
@@ -909,8 +909,7 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
     withSQLConf(SQLConf.ENABLE_DEFAULT_COLUMNS.key -> "false") {
       testDefaultColumn
     }
-    withSQLConf(SQLConf.ENABLE_DEFAULT_COLUMNS.key -> "true",
-      SQLConf.USE_NULLS_FOR_MISSING_DEFAULT_COLUMN_VALUES.key -> "false") {
```
Comment (gengliangwang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala:913]: Let's keep this `withSQLConf`, in case we change the default value in the future release.

**Review by gengliangwang — APPROVED** [review]
LGTM except minor comments.
Thanks for the work

**Comment by morvenhuang** [comment]
@gengliangwang thanks a lot for your time, I've put those 3 SQLConf back, thank you.

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by dtenedor** [comment]
@morvenhuang @gengliangwang there might be a bug in this. Should this:

```
      case SubqueryAlias(_, r: View) if r.isTempView =>
        StructType(r.schema.fields.dropRight(
          enclosingInsert.get.partitionSpec.size))
```

really be this?

```
      case SubqueryAlias(_, r: View) if r.isTempView =>
        StructType(r.desc.schema.fields.dropRight(
          enclosingInsert.get.partitionSpec.size))
```

The `r.schema` may trigger an exception if the view is not resolved, whereas `r.desc.schema` should always work. Should we consider a follow up patch to fix?

image

**Comment by morvenhuang** [comment]
@dtenedor sorry fo the late. I see that ResolveRelations rule is invoked before ResolveDefaultColumns, does that mean we'll never encounter unresolved table or view in ResolveDefaultColumns? Could you be kind enough to explain a bit more about this, so that I can have a better understanding.

**Comment by dtenedor** [comment]
@morvenhuang I'm moving `ResolveDefaultColumns` before `ResolveRelations` in a future pending PR: https://github.com/apache/spark/pull/36445. That change is needed to avoid throwing exceptions when processing MERGE commands before the current position of `ResolveDefaultColumns` in the main resolution rule batch.

If it's convenient, I can just make the update from `r.schema.fields` to `r.desc.schema.fields` in the same PR, that way it's less work for everyone.

**Comment by morvenhuang** [comment]
@dtenedor Oh, I see. It would be great if you can help to make that update in your new PR, thank you Daniel.
